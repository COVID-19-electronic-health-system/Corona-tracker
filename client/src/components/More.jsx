/* eslint-disable no-console */
/* eslint-disable no-alert */

import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogTitle from '@material-ui/core/DialogTitle';
import { Link, useHistory } from 'react-router-dom';
import axios from 'axios';
import { makeStyles } from '@material-ui/core/styles';
import MoreHorizIcon from '@material-ui/icons/MoreHoriz';
import { useBlockstack } from 'react-blockstack';
import { useTranslation } from 'react-i18next';
import { DialogContent, DialogContentText, TextField } from '@material-ui/core';
import buttonsCss from '../css/buttons';

const useStyle = makeStyles({
  root: {
    width: '100px',
  },
  dialog: {
    width: '100vw',
    textAlign: 'center',
  },
  text: {
    backgroundImage: 'linear-gradient(40deg, #4760ff 0%, #82a4f8 100%)',
    color: 'white',
  },
  dialogContent: {
    backgroundImage: 'linear-gradient(40deg, #7a9cf9 0%, #97b9f7 100%)',
  },
  descriptionText: {
    color: 'white',
    textAlign: 'left',
  },
  subscribeContainer: {
    align: 'center',
  },
  subtitleText: {
    color: 'white',
    margin: '0 auto',
  },
  action: {
    display: 'flex',
    flexDirection: 'column',
    backgroundImage: 'linear-gradient(#d7e1fa, #bbcef9)',
  },
  icon: {
    fontSize: '40px',
    color: 'white',
    transition: 'all .3s ease-in-out',
    '&:hover': {
      transform: 'scale(1.1)',
    },
  },
  buttons: {
    ...buttonsCss.buttons,
    width: '33vw',
    height: '8vh',
  },
  link: {
    height: '100%',
    width: '100vw',
    textDecoration: 'none',
    color: 'white',
    alignContent: 'center',
  },
});

const More = () => {
  const [open, setOpen] = useState(false);
  const [number, setNumber] = useState('');
  const classes = useStyle();
  const { signOut } = useBlockstack();
  const { t } = useTranslation();
  const history = useHistory();
  const handleClickOpen = () => {
    setOpen(!open);
  };
  const handleClose = () => {
    setOpen(false);
  };

  const subscribe = () => {
    axios
      .post(
        'https://m72j7fayxh.execute-api.us-east-1.amazonaws.com/api/subscribe',
        {
          number,
        },
        {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
        }
      )
      .then(res => {
        console.log(res);
        alert(
          `Subscribed successfully! You will be automatically unsubscribed in one day, and will receive three texts in that timespan. If you enter again, you will receive double the notifications - so please do not! This is very early alpha :)`
        );
      })
      .catch(err => {
        console.log(err);
      });
  };

  const navigateTo = href => {
    history.push(href);
  }

  return (
    <Link to="#more" className={classes.root} onClick={handleClickOpen}>
      <MoreHorizIcon className={classes.icon} />
      <Dialog
        className={classes.dialog}
        open={open}
        onClose={handleClose}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title" className={classes.text}>
          More
        </DialogTitle>
        <DialogContent className={classes.dialogContent}>
          <DialogContent>
            <DialogContentText className={classes.descriptionText}>
              Enter your phone number to be subscribed to the occasional question/survey to answer over text
            </DialogContentText>
            <TextField
              className={classes.subtitleText}
              onChange={e => setNumber(e.target.value)}
              autoFocus
              margin="dense"
              id="filled-phone"
              label={t('phoneNumber')}
              type="email"
              fullWidth
            />
          </DialogContent>
          <DialogActions className={classes.subscribeContainer}>
            <Button onClick={() => subscribe()} color="primary" className={classes.subtitleText}>
              Subscribe
            </Button>
          </DialogActions>
          <DialogActions>
            <Button size="medium" onClick={() => navigateTo('/onboard')} variant="contained" className={classes.buttons}>
              Settings
            </Button>
            <Button size="medium" onClick={() => navigateTo('/about')} variant="contained" className={classes.buttons}>
              About
            </Button>
            <Button size="medium" color="secondary" variant="contained" className={classes.buttons} onClick={signOut}>
              {t('signoutButtonText')}
            </Button>
          </DialogActions>
        </DialogContent>
      </Dialog>
    </Link>
  );
};

export default More;
