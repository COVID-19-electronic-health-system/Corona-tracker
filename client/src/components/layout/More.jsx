import React from 'react';
import { makeStyles } from '@material-ui/core/styles';

import PropTypes from 'prop-types';
import { Drawer, Typography, Grid, Button } from '@material-ui/core';
import FavoriteBorderIcon from '@material-ui/icons/FavoriteBorder';
import WavesIcon from '@material-ui/icons/Waves';
import actions from 'redux/actions/actions';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

const useStyles = makeStyles(theme => ({
  paper: {
    borderRadius: theme.spacing(1),
    width: theme.spacing(32),
    height: theme.spacing(14),
    border: 0,
    margin: 'auto',
    marginBottom: theme.spacing(9),

    background: `linear-gradient(0deg, #4760ff, #82a4f8)`,
  },
  text: {
    color: 'white',
    marginTop: '10px',
  },
  pulse: {
    fontSize: '40px',
  },
  spirometer: {
    fontSize: '40px',
    width: '60px',
  },
  buttonSpirometer: {
    color: 'white',
    marginRight: '30px',
  },
  buttonPulse: {
    color: 'white',
    marginLeft: '30px',
  },
}));

const More = props => {
  const { moreToggle, setMoreToggle } = props;
  const classes = useStyles();
  const handleClose = () => {
    setMoreToggle(!moreToggle);
  };
  return (
    <div>
      <Drawer
        open={moreToggle}
        anchor="bottom"
        onClose={handleClose}
        transitionDuration={500}
        BackdropProps={{ invisible: true }}
        classes={{ paper: classes.paper }}
      >
        <Grid container justify="center">
          <Grid item xs={12}>
            <Typography align="center" variant="body2" className={classes.text}>
              Explor More Open Sources Softwares
            </Typography>
          </Grid>
          <Button className={classes.buttonSpirometer}>
            <Grid container direction="column">
              <WavesIcon className={classes.spirometer} onClick={handleClose} />
              <Typography variant="caption" align="center">
                Spirometer
              </Typography>
            </Grid>
          </Button>
          <Button component={Link} to="/pulse" className={classes.buttonPulse} onClick={handleClose}>
            <Grid container direction="column" alignContent="center">
              <FavoriteBorderIcon className={classes.pulse} />
              <Typography variant="caption" align="center">
                Pulse
              </Typography>
            </Grid>
          </Button>
        </Grid>
      </Drawer>
    </div>
  );
};
More.propTypes = {
  setMoreToggle: PropTypes.func.isRequired,
  moreToggle: PropTypes.bool.isRequired,
};

const mapStateToProps = state => {
  return {
    moreToggle: state.navToggle.moreToggle,
    settingsToggle: state.navToggle.settingsToggle,
  };
};
const mapDispatchToProps = dispatch => {
  return {
    setMoreToggle: moreToggle => dispatch(actions.setMoreToggle(moreToggle)),
    setSettingsToggle: settingsToggle => dispatch(actions.setSettingsToggle(settingsToggle)),
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(More);
