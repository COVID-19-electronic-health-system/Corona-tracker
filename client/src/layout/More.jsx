/* eslint-disable no-console */
/* eslint-disable no-alert */
/* eslint-disable no-useless-escape */

import React, { useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogTitle from '@material-ui/core/DialogTitle';

import { makeStyles } from '@material-ui/core/styles';
import { useBlockstack } from 'react-blockstack';
import { useTranslation } from 'react-i18next';
import { DialogContent, DialogContentText, TextField, Snackbar } from '@material-ui/core';
import MuiAlert from '@material-ui/lab/Alert';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import buttonsCss from 'styles/buttons';
import actions from 'redux/actions/actions';

const useStyle = makeStyles(theme => ({
  root: {
    width: '80px',
    marginTop: '8px',
    cursor: 'pointer',
    [theme.breakpoints.up('md')]: {
      width: '93px',
    },
  },
  grow: {
    color: 'white',
    transition: 'all .3s ease-in-out',
    '&:hover': {
      transform: 'scale(1.2)',
    },
  },
  dialog: {
    width: '100vw',
    textAlign: 'center',
  },
  text: {
    backgroundImage: 'linear-gradient(40deg, #4760ff 0%, #82a4f8 100%)',
    color: 'white',
  },
  dialogContent: {
    backgroundImage: 'linear-gradient(40deg, #7a9cf9 0%, #97b9f7 100%)',
  },
  descriptionText: {
    textDecoration: 'none',
    color: 'white',
    textAlign: 'center',
    marginTop: '10px',
  },
  subscribeContainer: {
    align: 'center',
  },
  subtitleText: {
    marginTop: '1em',
    color: 'white',
    margin: '0 auto',
  },
  action: {
    display: 'flex',
    flexDirection: 'column',
    backgroundImage: 'linear-gradient(#d7e1fa, #bbcef9)',
  },
  icon: {
    fontSize: '40px',
  },
  buttons: {
    ...buttonsCss.buttons,
    width: '33vw',
    height: '8vh',
    minHeight: '50px',
  },
  link: {
    height: '100%',
    width: '100vw',
    textDecoration: 'none',
    color: 'white',
    alignContent: 'center',
  },
}));

function Alert(props) {
  const { severity, children, onClose } = props;
  return (
    <MuiAlert elevation={6} variant="filled" severity={severity} onClose={onClose}>
      {children}
    </MuiAlert>
  );
}

Alert.propTypes = {
  severity: PropTypes.string,
  children: PropTypes.node,
  onClose: PropTypes.func.isRequired,
};

Alert.defaultProps = {
  severity: '',
  children: '',
};

const More = props => {
  const {
    setReminderStatus,
    setSubscribedNumber,
    unsubscribeNumber,
    subscribedNumber,
    clearResponse,
    error,
    success,
    setMoreToogle,
    moreToggel,
  } = props;
  const [openSnackbar, setOpenSnackbar] = useState(false);
  const [snackbar, setSnackbar] = useState({});
  const [phoneNumber, setPhoneNumber] = useState('');
  const classes = useStyle();
  const { signOut, userSession } = useBlockstack();
  const { t } = useTranslation();

  const handleCloseSnackbar = (event, reason) => {
    if (reason === 'clickaway') {
      return;
    }
    if (success || error.response) {
      clearResponse();
    }
    setOpenSnackbar(false);
  };

  const handleSubmitSubscribe = () => {
    if (!/^\+[1-9]\d{1,14}$/.test(phoneNumber)) {
      setSnackbar({
        severity: 'error',
        message: 'Invalid phone number. Please try again',
      });
      setOpenSnackbar(true);
    } else {
      setSubscribedNumber(userSession, phoneNumber);
    }
    setPhoneNumber('');
  };

  useEffect(() => {
    let message;
    let severity;
    if (error.response) {
      severity = 'error';
      if (error.response.status === 409) {
        message = 'This phone number was already subscribed';
      } else {
        message = 'Something went wrong.  Please try again.';
      }
    }
    if (success) {
      severity = 'success';
      message =
        success === 'subscribed'
          ? `Subscribed successfully! You will be automatically unsubscribed in one day, and will receive three texts in that timespan. If you enter again, you will receive double the notifications - so please do not! This is very early alpha :)`
          : 'Unsubscribed successfully';
    }
    setSnackbar({
      message,
      severity,
    });
  }, [error, success]);

  return (
    <div>
      <Dialog
        className={classes.dialog}
        open={moreToggel}
        onClose={() => setMoreToogle(!moreToggel)}
        aria-labelledby="alert-dialog-title"
        aria-describedby="alert-dialog-description"
      >
        <DialogTitle id="alert-dialog-title" className={classes.text}>
          More
        </DialogTitle>
        <DialogContent className={classes.dialogContent}>
          <DialogContent>
            <DialogContentText className={classes.descriptionText}>
              Enter your phone number to subscribe/unsubscribe to the occasional question/survey to answer over text.
            </DialogContentText>
            <a
              className={classes.descriptionText}
              href="https://support.twilio.com/hc/en-us/articles/223183008-Formatting-International-Phone-Numbers"
            >
              Please add your +country code before entering
            </a>
            {subscribedNumber ? (
              <DialogContentText className={classes.descriptionText}>
                {`You are subscribed to text alerts at phone number: ${subscribedNumber}`}
              </DialogContentText>
            ) : (
              <TextField
                className={classes.subtitleText}
                onChange={e => setPhoneNumber(e.target.value)}
                autoFocus
                margin="dense"
                id="filled-phone"
                label={t('phoneNumber')}
                type="email"
                fullWidth
              />
            )}
          </DialogContent>
          <DialogActions className={classes.subscribeContainer}>
            {!subscribedNumber && (
              <Button
                onClick={() => {
                  handleSubmitSubscribe();
                }}
                color="primary"
                className={classes.subtitleText}
              >
                Subscribe
              </Button>
            )}
            {subscribedNumber && (
              <Button
                onClick={() => {
                  unsubscribeNumber(userSession, subscribedNumber);
                }}
                color="secondary"
                className={classes.subtitleText}
              >
                Unsubscribe
              </Button>
            )}
          </DialogActions>
          <DialogActions>
            <Button
              component={Link}
              to="/settings"
              size="medium"
              onClick={() => {
                setMoreToogle(!moreToggel);
              }}
              variant="contained"
              className={classes.buttons}
            >
              Settings
            </Button>
            <Button
              component={Link}
              to="/about"
              size="medium"
              onClick={() => {
                setMoreToogle(!moreToggel);
              }}
              variant="contained"
              className={classes.buttons}
            >
              About
            </Button>
            <Button
              size="medium"
              color="secondary"
              variant="contained"
              className={classes.buttons}
              onClick={() => {
                setMoreToogle(!moreToggel);
                setReminderStatus(true);
                signOut();
              }}
            >
              {t('moreSection.text.buttonSignOut.signOut')}
            </Button>
          </DialogActions>
        </DialogContent>
        <Snackbar
          open={openSnackbar || success || error.response}
          autoHideDuration={6000}
          anchorOrigin={{ vertical: 'top', horizontal: 'center' }}
          onClose={handleCloseSnackbar}
        >
          <Alert onClose={handleCloseSnackbar} severity={snackbar.severity}>
            {snackbar.message}
          </Alert>
        </Snackbar>
      </Dialog>
    </div>
  );
};

More.propTypes = {
  setSubscribedNumber: PropTypes.func.isRequired,
  unsubscribeNumber: PropTypes.func.isRequired,
  error: PropTypes.shape({
    response: PropTypes.shape({
      status: PropTypes.number,
    }),
  }).isRequired,
  success: PropTypes.string.isRequired,
  clearResponse: PropTypes.func.isRequired,
  subscribedNumber: PropTypes.string,
  setReminderStatus: PropTypes.func.isRequired,
  setMoreToogle: PropTypes.func.isRequired,
  moreToggel: PropTypes.bool.isRequired,
};

More.defaultProps = {
  subscribedNumber: null,
};

const mapState = state => {
  return {
    subscribedNumber: state.onboardingReducer.phoneNumber.subscribedNumber,
    error: state.onboardingReducer.phoneNumber.error,
    success: state.onboardingReducer.phoneNumber.success,
    moreToggel: state.navToggle.moreToggel,
  };
};

const mapDispatch = dispatch => {
  return {
    setSubscribedNumber: (userSession, number) => dispatch(actions.setSubscribedNumber(userSession, number)),
    unsubscribeNumber: (userSession, number) => dispatch(actions.unsubscribeNumber(userSession, number)),
    clearResponse: () => dispatch(actions.clearResponse()),
    setReminderStatus: status => dispatch(actions.setReminderStatus(status)),
    setMoreToogle: moreToggel => dispatch(actions.setMoreToogle(moreToggel)),
  };
};

export default connect(mapState, mapDispatch)(More);
