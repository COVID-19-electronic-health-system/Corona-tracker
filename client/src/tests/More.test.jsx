import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import More from '../components/More';

test('renders More component', () => {
  render(
    <BrowserRouter>
      <More />
    </BrowserRouter>
  );
});

// TODO test button click subscribes someone (E2E, ensure alert pops up)
// TODO test about, settings, log out works
