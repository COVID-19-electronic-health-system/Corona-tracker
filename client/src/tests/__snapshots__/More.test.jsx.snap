// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders More component 1`] = `ReactWrapper {}`;
